certora/specs/CER-2-Operator/CER-68-Operator-authorization.spec

```diff
@@ -0,0 +1,142 @@
+
+    bytes19 addressPrefix;
+    address operator;
+    uint256 operatorBitField;
+
+    address ownerBefore = getOwnerOf(addressPrefix);
+
+    // if msg.sender is the currentContract, it means we are within permit() and
+    // we need to use executionContext.getOnBehalfOfAccount() instead.
+    address actualCaller = e.msg.sender;
+    if (e.msg.sender == currentContract) {
+        actualCaller = getExecutionContextOnBehalfOfAccount(e);
+    }
+
+    // call the setOperator() method.
+    setOperator(e, addressPrefix, operator, operatorBitField);
+
+    // sender is from the prefix itself and thus plausible to be the owner
+    assert(getAddressPrefix(actualCaller) == addressPrefix);
+    // the owner before the call was either not set or actualCaller already
+    assert(ownerBefore == 0 || ownerBefore == actualCaller);
+    // sender is stored as the owner of the address prefix
+    assert(actualCaller == getOwnerOf(addressPrefix));
+    // make sure the right bitfield was set
+    assert(getOperator(addressPrefix, operator) == operatorBitField);
+}
+
+rule onlyOwnerOrOperatorCanCallSetAccountOperator() {
+    env e;
+    address account;
+    address operator;
+    bool authorized;
+
+    // if msg.sender is the currentContract, it means we are within permit() and
+    // we need to use executionContext.getOnBehalfOfAccount() instead.
+    address actualCaller = e.msg.sender;
+    if (e.msg.sender == currentContract) {
+        actualCaller = getExecutionContextOnBehalfOfAccount(e);
+    }
+
+    // call the setAccountOperator method.
+    setAccountOperator(e, account, operator, authorized);
+
+    address owner = haveCommonOwner(account, actualCaller) ? actualCaller : getAccountOwner(e, account);
+
+    // Since setAccountOperator did not revert, the actualCaller
+    // must either be the owner or operator
+    assert(actualCaller == owner || actualCaller == operator);
+
+}
+
+// a copy of the internal ownerLookup
+ghost mapping(bytes19 => address) ownerLookupGhost {
+    init_state axiom forall bytes19 prefix. ownerLookupGhost[prefix] == 0;
+}
+// makes sure the ownerLookupGhost is updated properly
+hook Sstore EthereumVaultConnectorHarness.ownerLookup[KEY bytes19 prefix] address value STORAGE {
+    ownerLookupGhost[prefix] = value;
+}
+// makes sure that reads from ownerLookup after havocs are correct
+hook Sload address value EthereumVaultConnectorHarness.ownerLookup[KEY bytes19 prefix] STORAGE {
+    require(ownerLookupGhost[prefix] == value);
+}
+
+// check that an owner of a prefix is always from that prefix
+invariant OwnerIsFromPrefix(bytes19 prefix)
+    getOwnerOf(prefix) == 0 || getAddressPrefix(getOwnerOf(prefix)) == prefix
+    filtered { 
+        f -> !isMustRevertFunction(f) &&
+        // We can't handle these functions since they have `CALL`s in them
+        f.selector != sig:batch(IEVC.BatchItem[] calldata).selector
+        && 
+        f.selector != sig:call(address, address, uint256, bytes calldata).selector
```

Comment:
maybe you want to use the ghost and make it persistent. then you can prove on the batch and call function given the safely assumption that not calling to self?

Reply:
This seems reasonable, but this was not enough to get it working: https://prover.certora.com/output/65266/ef3d4b111db245018586961fc4d099db?anonymousKey=8b2fe6c466b7959a17c88fd8ba307d24fbf499f6.

What if I get this change in and then try to improve on the cases for `batch`/`call` later?

Reply:
Also tried using a preserved block: https://prover.certora.com/output/65266/beae2d5b6d2143cb90ea38533cfc85e6?anonymousKey=a40bcffd8d155af32cf9cf6658ea50e9d0a874c5

Reply:
We fixed this now. Thanks @nd-certora